c10/core/impl/TorchDispatchModeTLS.h

```diff
@@ -6,20 +6,57 @@
 namespace c10 {
 namespace impl {
 
+enum class TorchDispatchModeKey : int8_t {
+  PROXY = 0,
+  FAKE = 1,
+  NUM_MODE_KEYS = 2
+};
+
 struct C10_API TorchDispatchModeTLS {
+  // This API is NOT invariant safe.
+  // If you're pushing an infra mode onto the stack, we expect
+  // you to use set_mode
   static void push_onto_stack(std::shared_ptr<SafePyObject> mode);
+  // Pops the top mode of the stack,
+  // giving precedence to user modes before attempting to pop
+  // any infra modes
   static const std::shared_ptr<SafePyObject> pop_stack();
+  // Returns the highest-priority infra mode on the stack,
+  // along with its mode key.
+  static const std::tuple<std::shared_ptr<SafePyObject>, TorchDispatchModeKey>
+  pop_highest_infra_mode();
```

Comment:
It's not obvious to me why this function is useful

Reply:
The only actual use of this function today is `StashTorchDispatchModeGuard`, which is used in our dispatch logic. When we need to temporarily pop the highest mode on the stack (and push it back on exit), there are two cases:

(1) highest mode is a user mode: we can use `push_non_infra_mode_onto_stack()` on exit

(2) highest mode is an infra mode: we can't just blindly push the infra mode back on exit. We need to know its mode key, so we can call `set_mode()` accordingly.

Technically though... we could `getattr(mode "_mode_key")` from inside this C++ guard to just get the source-of-truth mode key, instead of asking the TLS what mode key corresponds to this stack. Then I could consolidate and use `pop_stack()`, and kill this method. WDYT?